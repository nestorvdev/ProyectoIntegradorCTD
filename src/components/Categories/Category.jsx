import React from 'react';
import Styles from './styles.module.css';


export default function Category({ imageUrl, title, description, category, handleCategory}) {

    return (
        <div className={Styles.categoryCard} onClick={() => {
            handleCategory(category)
          }}  >
            <img className={Styles.categoryImage} src={imageUrl} alt="imagen 1" />
            <div className={Styles.categoryType}>
                <h3>{title}</h3>
                <p>{description}</p>
            </div>
        </div >

    )
}

