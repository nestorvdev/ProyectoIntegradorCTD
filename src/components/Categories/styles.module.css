.categoryContainer {
  background-color: white; 
  width: 100%;
}

.categoryBlock{
  padding-bottom:5px;
  padding-left: clamp(10px,3.9%,40px);
  padding-right: clamp(10px,3.9%,40px);
  width: 100%;
}

.categoryBox {
  color: var(--secondary-color);
  display: flex;
  justify-content: space-between;
}

.categoryCard {
  border-radius: 10px;
  box-shadow: 0 4px 4px 0px rgba(0, 0, 0, 0.25);
  margin: 7px 0;
  max-width: 310px;
  width: 22.7vw;
  transition: all 0.3s;
}

.categoryBox>div:first-child{
  margin-left: 0px;;
}
.categoryBox>div:last-child{
  margin-right: 0px;;
}

.categoryCard:hover {
  box-shadow: 0 4px 4px 0px rgba(0, 0, 0, 0.5);
  transform: scale(1.02);
  cursor: pointer;
}

.categoryCard:active {  
  transform: scale(0.95);
}

.categoryImage {
  background-size: cover;
  border-radius: 10px 10px 0px 0px;
  height: 183.81px;
  max-width: 310px;
  width: 100%;
}

.image1 {
  background-image: url("./img/hoteles.jpg");
}

.image2 {
  background-image: url("./img/hostels.jpg");
}

.image3 {
  background-image: url("./img/departamentos.jpeg");
}

.image4 {
  background-image: url("./img/bedAndBreakfast.jpg");
}

.categoryBlock h2 {
  color: var(--secondary-color);
  font-size: var(--heading2);
  font-weight: var(--font-bold);
  margin-bottom: 9px;
  margin-left: 7px;
  padding-top: 16px;
}

.categoryType {
  display: flex;
  flex-direction: column;
  height: 63.7px;
  justify-content: center;
  margin-left: 11px;
  margin-right: 11px;
}

.categoryType h3 {
  font-size: var(--heading3);
  font-weight: var(--font-bold);
  overflow: hidden;
  white-space: nowrap;
}

.categoryType p {
  font-size: var(--text2);
}

@media all and (max-width: 769px) {
  .categoryBox {
    flex-wrap: wrap;
  }

  .categoryCard {
    max-width: 347px;
    width: 43vw;
  }

  .categoryImage {
    height: 181.3px;
    max-width: 347px;
  }

  .categoryType {
    height: 62.7px;
  }

  .categoryBlock h2 {
    overflow: hidden;
    white-space: nowrap;
  }
}

@media all and (max-width: 415px) { 
  .categoryBox {
    justify-content: center;
  }
  .categoryCard {
    margin: 7px 0;
    max-width: 392px;
    width: 94.7vw;
  }

  .categoryImage {
    height: 205.1px;
    max-width: 392px;
    width: 94.7vw;
  }

  .categoryType {
    height: 70.9px;
  }

  .categoryBlock h2 {
    font-size: var(--heading3);
    margin: 36px 0 3px 0;
  }
}
