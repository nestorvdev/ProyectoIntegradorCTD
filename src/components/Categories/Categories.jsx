import React from "react";
import Styles from "./styles.module.css";
import StylesApp from "../../App.module.css";
import Category from "./Category";
import axios from "axios";
import { useState, useEffect } from "react";
import Spinner from "../spinner/Spinner";

export default function Categories({ handleCategory }) {
  const baseURL = "http://worldguestbooking.com.ar:8080/categories/all";
  const [data, setData] = useState([]);
  const [loading, setLoading] = useState(true);
  const [errorMessage, setErrorMessage] = useState("");

  useEffect(() => {
    axios
      .get(baseURL)
      .then((response) => {
        setData(response.data);
        setLoading(false);
      })
      .catch((error) => {
        setErrorMessage(error);
      });
  }, []);

  return (
    (errorMessage && loading) ?
      <div className={`${Styles.categoryContainer} ${StylesApp.delimiter}`}>
        <div className={`${Styles.categoryBlock} ${StylesApp.delimiterChild}`}>
          <h2 className={Styles.categoryBlock}>
            Categorías no disponibles - Falta la conexión con el Back
          </h2>
        </div>
      </div>
      :
      <div className={`${Styles.categoryContainer} ${StylesApp.delimiter}`}>
        <div className={`${Styles.categoryBlock} ${StylesApp.delimiterChild}`}>
          <h2>Buscar por tipo de alojamiento</h2>
          {loading ? (
            <Spinner />
            /* <p>Loading Data...</p> */
          ) : (
            <div className={Styles.categoryBox}>
              {data.map((c) => (
                <Category
                  key={c.id}
                  title={c.title}
                  imageUrl={c.url}
                  description={c.description}
                  category={c.title}
                  handleCategory={handleCategory}
                />
              ))}
            </div>
          )}
        </div>
      </div>
  );
}