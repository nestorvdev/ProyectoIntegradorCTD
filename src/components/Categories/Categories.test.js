import React from "react";
import { shallow, mount } from "enzyme";
import "@testing-library/jest-dom";
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { createSerializer } from 'enzyme-to-json';
import Category from "./Category";
import Categories from "./Categories";


Enzyme.configure({ adapter: new Adapter() });
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));

let wrapper;

describe("<Category />", () => {
    let props = {
        handleCategory: jest.fn(), category: "test", key: 1, title: "Hoteles", imageUrl: { url: "test" }, description: "test"
    };

    beforeEach(() => {
        wrapper = shallow(<Category {...props} />);
    });

    it("Renderiza el componente correctamente", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('Deberia disparar el evento handleCategory al hacer click', () => {
        wrapper = shallow(<Category {...props} />);

        wrapper.find('div.categoryCard').first().props().onClick()
        expect(props.handleCategory).toBeCalled();
    });

    it('Debería renderizar la imagen con la url que viene por props', () => {
        expect(wrapper.find('img').props().src).toBe(props.imageUrl);
    })

    it('Debería renderizar el componente h3 con el titulo title de props', () => {
        expect(wrapper.find('h3').text()).toBe(props.title);
    })

    it('Debería renderizar el componente p con la descripción description de props', () => {
        expect(wrapper.find('p').text()).toBe(props.description);
    })
});

describe("<Categories />", () => {
    let props = {
        category: "test",
        handleCategory: jest.fn(),
    };
   /*  let data = {
        data: {
            Category: [
                {
                    key: 1,
                    title: "Hoteles",
                    imageUrl: "test",
                    description: "test",
                    category: "test",
                    handleCategory: jest.fn()
                }

            ]
        }
    }
    let useEffect = jest.spyOn(React, 'useEffect');

    const mockUseEffect = () => {
        useEffect.mockImplementationOnce(f => f());
    };

    beforeEach(() => {
        wrapper = shallow(<Categories {...props} />);

    });

    it("Renderiza el componente correctamente", () => {
        expect(wrapper).toMatchSnapshot();

    });

    it('El hook useEffect debe devolver un response.data', () => {
        mockUseEffect();
        mockUseEffect();
        wrapper = mount(<Categories {...props} />);
        let axiosData = jest.fn().mockResolvedValue(data);
        expect(axiosData).toHaveBeenCalled();
    });

    it('Debería renderizar el componente Category', () => {
        expect(wrapper.find('Category')).toHaveLength(1);
    });

    it('Debería renderizar el componente h2 con el texto correspondiente', () => {
        expect(wrapper.find('h2').text()).toBe("Buscar por tipo de alojamiento");
    });  */

    let wrapper = mount(<Categories {...props}/>); //Se usa mount para testear el useEffect
    beforeEach(() => {
        wrapper = mount(<Categories {...props}/>);        
    })
    it("Deberia mostrar <Categories/> correctamente", () => {
        expect(wrapper).toMatchSnapshot();
    });


});
