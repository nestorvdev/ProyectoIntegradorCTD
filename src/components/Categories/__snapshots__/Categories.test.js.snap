// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Categories /> Deberia mostrar <Categories/> correctamente 1`] = `
<div
  className="categoryContainer delimiter"
>
  <div
    className="categoryBlock delimiterChild"
  >
    <h2>
      Buscar por tipo de alojamiento
    </h2>
    <div
      className="spinnerContainer"
    >
      <div
        className="spinner"
      />
    </div>
  </div>
</div>
`;

exports[`<Category /> Renderiza el componente correctamente 1`] = `
<div
  className="categoryCard"
  onClick={[Function]}
>
  <img
    alt="imagen 1"
    className="categoryImage"
    src={
      Object {
        "url": "test",
      }
    }
  />
  <div
    className="categoryType"
  >
    <h3>
      Hoteles
    </h3>
    <p>
      test
    </p>
  </div>
</div>
`;
